src/stream/src/executor/merge.rs

```diff
@@ -218,6 +251,57 @@ impl SelectReceivers {
             last_base: 0,
             actor_id,
             barrier: None,
+            untransferred_watermarks: BTreeMap::default(),
+        }
+    }
+
+    fn check_heap(&mut self, col_idx: usize) -> Option<Watermark> {
+        let mut watermark_to_transfer = None;
+        let col_data = self.untransferred_watermarks.get_mut(&col_idx).unwrap();
+        while !col_data.first_untransferred_watermarks.is_empty()
+            && col_data.first_untransferred_watermarks.len()
+                == self.upstreams.len() + self.blocks.len()
+        {
+            let Reverse((watermark, actor_id)) =
+                col_data.first_untransferred_watermarks.pop().unwrap();
+            watermark_to_transfer = Some(watermark);
```

Comment:
Can we also pop and eliminate all other watermarks with the same value?

Reply:
> pop and eliminate all other watermarks with the same value

resolved